### PR TITLE
`Development`: Improve error message regex efficiency

### PR DESCRIPTION
**Pull request by MarkusPaulsen**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
Current regular expressions used for finding time out exceptions in programming exercise are inefficient. They use `*` matches which are slow and venerable for false positives. 

### Description
This PR introduces optimised regular expressions for finding time out exceptions in programming exercise. New expressions use explicit symbols matchers and contain a match length limit. Java Patters representing them are created in the static scope, so their compilation does not slow down the server.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Students
- 1 Programming Exercise (Java)

1. Edit the tests of the exercise, so that at least one test simulated time out. One way to achieve it is by replacing test's contents with `throw new Exception("execution timed out after 1 s");`. Another way you be to enforce awaitability's `await()` exception.
2. Create any submission to this exercise.
3. Verify that the exception is parsed properly and time out message appears in the run report. The message is displayed in the screenshots section of this description.

#### Performance Review
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ProgrammingExerciseFeedbackCreationService.java | 86% | ✅                           |

### Screenshots

Time out error message:
image

## Summary by CodeRabbit

- **Refactor**
  - Improved timeout exception handling in the programming exercise feedback service for enhanced accuracy.

- **Tests**
  - Added new test case to ensure reliable feedback creation for general timeout scenarios.

@coderabbitai ignore

### CODE REVIEW

**Review by RY997 — APPROVED** [review]
Tested on TS2, worked as expected

**Review by egekurt123 — DISMISSED** [review]
Manually tested on ts5, works as expected

**Review by marlon-luca-bu — APPROVED** [review]
LGTM, works as expected

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java [hunk 1]
```diff
@@ -45,8 +45,24 @@ public class ProgrammingExerciseFeedbackCreationService {
 
     private static final Predicate<String> IS_PYTHON_EXCEPTION_LINE = line -> line.startsWith(PYTHON_EXCEPTION_LINE_PREFIX);
 
-    private static final List<String> TIMEOUT_EXCEPTIONS = Arrays.asList("org.junit.runners.model.TestTimedOutException", "java.util.concurrent.TimeoutException",
-            "org.awaitility.core.ConditionTimeoutException", "Timed?OutException");
+    private static final String TIMEOUT_EXCEPTIONS_REGEX = String.join("|", "org.junit.runners.model.TestTimedOutException", "java.util.concurrent.TimeoutException",
+            "org.awaitility.core.ConditionTimeoutException", "TimedOutException");
+
+    private static final String EXCEPTION_PREFIX_MESSAGE_REGEX = "[a-zA-Z0-9._ ]{1,5000}";
```
Comment (nityanandaz) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java:51]: `a-zA-Z_0-9` is also contained in the word character class. Would that

```
[\w. ]{1,5000}
```

work?

https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html

**Review by nityanandaz — COMMENTED** [review]
Code review

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java (hunk 1 above)
Comment (jakubriegel) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java]: Great idea! Thanks

**Review by nityanandaz — DISMISSED** [review]
Code review: good improvement :)

**Review by marlon-luca-bu — DISMISSED** [review]
Reapprove, after implementing improvements

**Review by egekurt123 — APPROVED** [review]
Reapprove

**Review by RY997 — APPROVED** [review]
Reapprove

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java [hunk 2]
```diff
@@ -45,8 +45,24 @@ public class ProgrammingExerciseFeedbackCreationService {
 
     private static final Predicate<String> IS_PYTHON_EXCEPTION_LINE = line -> line.startsWith(PYTHON_EXCEPTION_LINE_PREFIX);
 
-    private static final List<String> TIMEOUT_EXCEPTIONS = Arrays.asList("org.junit.runners.model.TestTimedOutException", "java.util.concurrent.TimeoutException",
-            "org.awaitility.core.ConditionTimeoutException", "Timed?OutException");
+    private static final String TIMEOUT_EXCEPTIONS_REGEX = String.join("|", "org.junit.runners.model.TestTimedOutException", "java.util.concurrent.TimeoutException",
+            "org.awaitility.core.ConditionTimeoutException", "TimedOutException");
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java:49]:
```suggestion
            "org.awaitility.core.ConditionTimeoutException", "Timed?OutException");
```
Difference compared to old variant. Does that matter? I assume `TimeOutException`s might exist somewhere.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java [hunk 3]
```diff
@@ -45,8 +45,24 @@ public class ProgrammingExerciseFeedbackCreationService {
...
+
+    private static final String EXCEPTION_PREFIX_MESSAGE_REGEX = "[\\w. ]{1,5000}";
```
[the remark this answers is not in the document]
Comment (MarkusPaulsen) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java:51]: @b-fein According to ChatGPT, the following symbols are included in ".", but not in [\w. ]:
- Newline Characters
- Whitespace Characters except space
- Control Characters
- Unicode Characters
- Special Characters (e.g. @, #, $, ...) except .

Can you think of any scenario, where a timeout exception would throw one of these characters?

[the remark this answers is not in the document]
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseFeedbackCreationService.java:51]: This is a `.` inside the character set, though. So it matches only a literal `.`. (unless in an example like `.*` where it matches _any_ character)

I could for example imagine that also `,-_` are part of the message. Not sure what awaitility adds, though.

**Comment by krusche** [comment]
There is still a valid comment by @b-fein that we should address

**Review by Gusti2010 — CHANGES_REQUESTED** [review]
After changing a test case to generate a timeout the build fails after pushing a solution. Did i misunderstood the test steps? (tested manually on ts11_Legacy)

test case:
Screenshot 2024-01-31 at 13 30 07

after pushing a solution:
Screenshot 2024-01-31 at 13 30 20

**Comment by krusche** [comment]
This PR has been open for too long and will be closed.

**Pull request closed without merging**